src/functional/ffront/func_to_foast.py

```diff
@@ -16,23 +16,153 @@
 import textwrap
 import types
-from typing import Optional
+from collections.abc import Callable, Iterator
+from dataclasses import dataclass
+from typing import Any, Optional, Union
 
+from eve import typingx
 from eve.type_definitions import SourceLocation
 from functional import common
+from functional.ffront import fbuiltins
 from functional.ffront import field_operator_ast as foast
+from functional.ffront import symbol_makers
 from functional.ffront.ast_passes import (
     SingleAssignTargetPass,
     SingleStaticAssignPass,
+    StringifyAnnotationsPass,
     UnpackedAssignPass,
 )
 from functional.ffront.foast_passes.type_deduction import FieldOperatorTypeDeduction
-from functional.ffront.type_parser import FieldOperatorTypeParser
 
 
+MISSING_FILENAME = "<string>"
+
+
+def _make_source_definition_from_function(func: Callable) -> SourceDefinition:
+    try:
+        filename = inspect.getabsfile(func) or MISSING_FILENAME
+        source = textwrap.dedent(inspect.getsource(func))
+        starting_line = (
+            inspect.getsourcelines(func)[1] if not filename.endswith(MISSING_FILENAME) else 1
+        )
+    except OSError as err:
+        if filename.endswith(MISSING_FILENAME):
+            message = "Can not create field operator from a function that is not in a source file!"
+        else:
+            message = f"Can not get source code of passed function ({func})"
+        raise ValueError(message) from err
+
+    return SourceDefinition(source, filename, starting_line)
+
+
+def _make_closure_refs_from_function(func: Callable) -> ClosureRefs:
+    (nonlocals, globals, inspect_builtins, inspect_unbound) = inspect.getclosurevars(  # noqa: A001
+        func
+    )
+    # python builtins returned by getclosurevars() are not ffront.builtins
+    unbound = set(inspect_builtins.keys()) | inspect_unbound
+    builtins = unbound & set(fbuiltins.ALL_BUILTIN_NAMES)
+    unbound -= builtins
+    annotations = typingx.get_type_hints(func)
+
+    return ClosureRefs(nonlocals, globals, annotations, builtins, unbound)
+
+
+def _make_symbol_names_from_source(source: str, filename: str = MISSING_FILENAME) -> SymbolNames:
+    try:
+        mod_st = symtable.symtable(source, filename, "exec")
+    except SyntaxError as err:
+        raise common.GTValueError(
+            f"Unexpected error when parsing provided source code (\n{source}\n)"
+        ) from err
+
+    assert mod_st.get_type() == "module"
+    if len(children := mod_st.get_children()) != 1:
+        raise common.GTValueError(
+            f"Sources with multiple function definitions are not yet supported (\n{source}\n)"
+        )
+
+    func_st = children[0]
+
+    param_names = set()
+    imported_names = set()
+    local_names = set()
+    for name in func_st.get_locals():
+        if (s := func_st.lookup(name)).is_imported():
+            imported_names.add(name)
+        elif s.is_parameter:
+            param_names.add(name)
+        else:
+            local_names.add()
```

Comment:
is this `.add()` empty on purpose? If so, why is it there?

Reply:
Good catch. It's a typo